lib/gems/pending/util/miq_file_storage.rb

```diff
@@ -0,0 +1,254 @@
+    klass = storage_interface_classes[scheme]
+
+    raise InvalidSchemeError, scheme if klass.nil?
+
+    klass.new_with_opts(opts)
+  end
+  private_class_method :fetch_interface_class
+
+  def self.storage_interface_classes
+    @storage_interface_classes ||= Interface.descendants.each_with_object({}) do |klass, memo|
+      memo[klass.uri_scheme] = klass if klass.uri_scheme
+    end
+  end
+
+  class Interface
+    BYTE_HASH_MATCH = /^(?<BYTE_NUM>\d+(\.\d+)?)\s*(?<BYTE_QUALIFIER>K|M|G)?$/i
+    BYTE_HASH       = {
+      "k" => 1.kilobyte,
+      "m" => 1.megabyte,
+      "g" => 1.gigabyte
+    }.freeze
+
+    attr_reader :remote_file_path, :byte_count, :source_input, :input_writer
+
+    def self.new_with_opts(opts) # rubocop:disable Lint/UnusedMethodArgument
+      raise NotImplementedError, "#{name}.new_with_opts is not defined"
+    end
+
+    def self.uri_scheme
+      nil
+    end
+
+    # :call-seq:
+    #   add( remote_uri )                       { |input_writer| ... }
+    #   add( remote_uri, byte_count )           { |input_writer| ... }
+    #
+    #   add( local_io, remote_uri )
+    #   add( local_io, remote_uri, byte_count )
+    #
+    # Add a file to the destination URI.
+    #
+    # In the block form of the method, only the remote_uri is required, and it
+    # is assumed the input will be a generated in the executed block (most
+    # likely an external process) to a unix pipe that can be written to.  The
+    # pipe generated by this method and passed in to the block as a file
+    # location to the `input_stream`).
+    #
+    # In the non-block form, a source must be provided as the first argument
+    # either as an IO object that can be read from, or a file path, and the
+    # second argument is the remote_uri as in the block form.
+    #
+    # An additional argument in both forms as the last argument is `byte_count`
+    # can also be included.  If passed, it will be assumed that the resulting
+    # input will be split, and the naming for the splits will be:
+    #
+    #   - filename.00001
+    #   - filename.00002
+    #   ...
+    #
+    # Block form:
+    #
+    #   nfs_session.add("path/to/file", "200M") do |input_stream|
+    #     `pg_dump -f #{input_stream} vmdb_production`
+    #   end
+    #
+    # Non-block form:
+    #
+    #   nfs_session.add("path/to/local_file", "path/to/remote_file")
+    #   nfs_session.add("path/to/local_file", "path/to/remote_file", "200M")
+    #
+    def add(*upload_args, &block)
+      initialize_upload_vars(*upload_args)
+      mkdir(File.dirname(@remote_file_path))
+      thread = handle_io_block(&block)
+      result = if byte_count
+                 upload_splits
+               else
+                 upload_single(@remote_file_path)
+               end
+      # `.join` will raise any errors from the thread, so we want to do that
+      # here (if a thread exists of course).
+      thread.join if block_given?
```

Comment:
I think this would be more clear if we used the thread's existence as the condition rather than if the block was provided.

It's a bit confusing that I have to look into `#handle_io_block` to figure out that a thread is only created if a block is provided.

Reply:
Sure, makes sense to me.  I think I was just using that as the check elsewhere, but I think that is a reasonable change.